src/main/webapp/view/discussion/discussion-comment.js

```diff
@@ -0,0 +1,152 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {timestampToString} from '../../timestamps.js';
+import {ELEMENT_DISCUSSION} from './discussion-area.js';
+import {COMMENT_TYPE_REPLY} from './discussion.js';
+
+/**
+ * Renders a comment and its replies, with a form to post a new reply.
+ */
+export default class DiscussionComment extends HTMLElement {
+  static #TEMPLATE = document.querySelector('#comment-template');
+
+  static #SLOT_HEADER = 'header';
+  static #SLOT_CONTENT = 'content';
+  static #SLOT_REPLIES = 'replies';
+
+  static #SELECTOR_SHOW_REPLY = '#show-reply';
+  static #SELECTOR_REPLY_FORM = '#reply-form';
+  static #SELECTOR_CANCEL_REPLY = '#cancel-reply';
+  static #SELECTOR_POST_REPLY = '#post-reply';
+  static #SELECTOR_REPLY_TEXTAREA = '#reply-textarea';
+
+  #discussion;
+
+  /**
+   * Creates an custom HTML element representing a comment.  This uses the
+   * template and slots defined by `TEMPLATE` to render the
+   * comment's content and replies, once the comment is set.
+   *
+   * @param {DiscussionArea} discussion The current discussion.
+   */
+  constructor(discussion) {
+    super();
+    this.#discussion = discussion;
+    this.attachShadow({mode: 'open'});
+    const shadow = DiscussionComment.#TEMPLATE.content.cloneNode(true);
+    this.shadowRoot.appendChild(shadow);
+    this.addReplyEventListeners();
+  }
+
+  /**
+   * @param comment The comment from the discussion that this element should
+   *     render.
+   */
+  setComment(comment) {
+    this.comment = comment;
+    this.textContent = '';
+    this.setSlotSpan(
+        DiscussionComment.#SLOT_HEADER, this.getHeaderString(comment));
+    this.setSlotSpan(DiscussionComment.#SLOT_CONTENT, comment.content);
+    this.addReplies(comment.replies);
+  }
+
+  /**
+   * Returns a string containing the timestamp, author, and creation time of
+   * the `comment`.  The timestamp is not displayed for replies to
+   * other comments.
+   */
+  getHeaderString(comment) {
+    const username = comment.author.email.split('@')[0];
+    let timestampPrefix = '';
+    if (comment.type !== COMMENT_TYPE_REPLY) {
+      // Don't show timestamp on replies.
+      timestampPrefix = `${timestampToString(comment.timestampMs.value)} - `;
+    }
+    return `${timestampPrefix}${username} on ${comment.created}`;
+  }
+
+  /**
+   * Adds event listeners and handlers for the reply button and form.
+   * This makes the form open when "Reply" is pressed, submit when
+   * "Post" is pressed, and close when "Cancel" is pressed.
+   */
+  addReplyEventListeners() {
```

Comment:
Can this be marked static so it is clear in the constructor we aren't playing with the internal state?

Reply:
You can pass in the shadowRoot.

Reply:
Talked offline, and this change would be too complicated--keeping as-is